Interaction regions should support occlusions

etiennesegonzac opened the pull request:
#### 0c8fb49c7a40115abe543963829eb8268b96d048
<pre>
Interaction regions should support occlusions
<a href="https://bugs.webkit.org/show_bug.cgi?id=250860">https://bugs.webkit.org/show_bug.cgi?id=250860</a>
&lt;rdar://103928196&gt;

Reviewed by Tim Horton.

Maintain a separate layer tree for Interaction Regions on top of the
content. This &quot;mirror layer tree&quot; includes layers for occlusions and now
comes with clear ordering rules.

* LayoutTests/interaction-region/click-handler-dynamically-added-expected.txt:
* LayoutTests/interaction-region/click-handler-expected.txt:
* LayoutTests/interaction-region/click-handler-in-shadowed-layer-expected.txt:
* LayoutTests/interaction-region/event-region-overflow-expected.txt:
* LayoutTests/interaction-region/icon-inside-button-single-region-expected.txt:
* LayoutTests/interaction-region/inline-link-dark-background-expected.txt:
* LayoutTests/interaction-region/inline-link-expected.txt:
* LayoutTests/interaction-region/inline-link-in-composited-iframe-expected.txt:
* LayoutTests/interaction-region/inline-link-in-layer-expected.txt:
* LayoutTests/interaction-region/inline-link-in-non-composited-iframe-expected.txt:
* LayoutTests/interaction-region/inline-link-with-pointer-events-none-content-expected.txt:
* LayoutTests/interaction-region/input-type-file-region-expected.txt:
* LayoutTests/interaction-region/input-type-range-region-expected.txt:
* LayoutTests/interaction-region/paused-video-regions-expected.txt:
* LayoutTests/interaction-region/split-inline-link-expected.txt:
* LayoutTests/interaction-region/wrapped-inline-link-expected.txt:
Update expectations with the new dump format.

* LayoutTests/interaction-region/layer-tree-expected.txt: Added.
* LayoutTests/interaction-region/layer-tree.html: Added.
Add test covering the final RemoteLayerTree structure.

* LayoutTests/interaction-region/overlay-expected.txt: Renamed from LayoutTests/interaction-region/inline-link-in-aria-hidden-subtree-expected.txt.
* LayoutTests/interaction-region/overlay.html: Renamed from LayoutTests/interaction-region/inline-link-in-aria-hidden-subtree.html.
Remove the aria-hidden test.
Add test for the occlusions detection.

* LayoutTests/interaction-region/region-area-overflow-does-not-crash-expected.txt:
* LayoutTests/interaction-region/region-area-overflow-does-not-crash.html:
Update test to also include a big occlusion region.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.getCALayerTree):
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::caLayerTreeAsText const):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::caLayerTreeAsText const):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _caLayerTreeAsText]):
(dumpCALayer):
Introduce a new test helper to get the full remote layer hierarchy.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
Remove the aria-hidden check. It was too strict and is not needed
anymore.
Generate regions for elements likely to occlude interactions.
(WebCore::operator&lt;&lt;):
Update the dump format to differentiate interactions and occlusions.

* Source/WebCore/page/InteractionRegion.h:
(WebCore::operator==):
(WebCore::InteractionRegion::encode const):
(WebCore::InteractionRegion::decode):
Add a `type` enum to the InteractionRegion struct to differentiate
interactions and occlusions.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::applyGeometryPropertiesToLayer):
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
Extract the geometry related properties application code to a new
method.
(WebKit::RemoteLayerTreePropertyApplier::applyProperties):
Only update geometry and event region properties for Interaction Regions
layers.
(WebKit::applyInteractionRegionsHierarchyUpdate):
(WebKit::RemoteLayerTreePropertyApplier::applyHierarchyUpdates):
Apply all hierarchy updates to the Interaction Regions mirror layer tree.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateLayerTree):
Keep the Interaction Regions mirror layer tree on top of the content after
each layer tree update.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::isInteractionLayer):
(WebKit::isOcclusionLayer):
(WebKit::isAnyInteractionRegionLayer):
(WebKit::isInteractionRegionLayer): Deleted.
(WebKit::setInteractionRegionOcclusion):
Update utility functions to differentiate interactions and occlusions.
(WebKit::insertInteractionRegionLayersForLayer):
(WebKit::appendInteractionRegionLayersForLayer): Deleted.
Insert the Interaction Regions back at the beginning of the sublayers array.
(WebKit::updateLayersForInteractionRegions):
Maintain the following order when updating Interaction Regions layers:
1. Occlusions always come first (in no particular order).
2. Interactions come next, ordered by area (biggest to smallest).
3. Other siblings come last, in the same order as the content layers they mirror.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::interactionRegionsLayer const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::detachFromParent):
(WebKit::RemoteLayerTreeNode::initializeLayer):
Add an `interactionRegionsLayer` property to the RemoteLayerTreeNode.

Canonical link: <a href="https://commits.webkit.org/259504@main">https://commits.webkit.org/259504@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a63f6dc94d981bf1d442ffa4648a9bcec208d31d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104524 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13602 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37427 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113798 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174022 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108444 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14710 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4525 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96859 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112756 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110291 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11359 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94397 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38928 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93206 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26009 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80598 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6958 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27366 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7078 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3942 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13113 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46926 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6550 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8869 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

etiennesegonzac commented on LayoutTests/interaction-region/layer-tree-expected.txt:
```diff
@@ -0,0 +1,132 @@
+
```
Note: this test is more deterministic than the others, because the remote layer tree is ordered.
If the `UIHelper.getCALayerTree()` thing makes sense, we could upgrade the other tests too in a separate patch.

hortont424 commented on Source/WebCore/page/InteractionRegion.h:
```diff
@@ -48,6 +48,7 @@ struct InteractionRegion {
     ElementIdentifier elementIdentifier;
     Region regionInLayerCoordinates;
     float borderRadius { 0 };
+    bool isOccluding { false };
```
This should probably be an `enum class X : bool` so we can have good names at initialization sites instead of random true/false in the last position

hortont424 commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -163,19 +160,10 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
-    if (properties.changedProperties & RemoteLayerTreeTransaction::BorderColorChanged)
-        layer.borderColor = cgColorFromColor(properties.borderColor).get();
 
     if (properties.changedProperties & RemoteLayerTreeTransaction::BorderWidthChanged)
```
vaguely weird that border-width is a "geometry" property

hortont424 commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
```diff
@@ -188,23 +176,40 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
-    if (properties.changedProperties & RemoteLayerTreeTransaction::DoubleSidedChanged)
-        layer.doubleSided = properties.doubleSided;
-
     if (properties.changedProperties & RemoteLayerTreeTransaction::MasksToBoundsChanged)
```
This one too. Also not geometry

hortont424 commented on Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
```diff
@@ -121,6 +121,7 @@ class UIScriptControllerIOS final : public UIScriptControllerCocoa {
     JSObjectRef inputViewBounds() const override;
     JSRetainPtr<JSStringRef> scrollingTreeAsText() const override;
     JSRetainPtr<JSStringRef> uiViewTreeAsText() const override;
+    JSRetainPtr<JSStringRef> caLayerTreeAsText() const override;
```
I think this should go in UIScriptControllerCocoa -- CA is identical on both platforms.

hortont424 commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -139,11 +136,25 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
     // FIXME: Consider also allowing elements that only receive touch events.
-    if (!renderer.style().eventListenerRegionTypes().contains(EventListenerRegionType::MouseClick))
+    auto cursor = cursorTypeForElement(*element);
+    if (!renderer.style().eventListenerRegionTypes().contains(EventListenerRegionType::MouseClick)
```
Maybe give this set of conditions a local with a meaningful name so it's easier to follow that this is the "non-interactive" path?

etiennesegonzac committed:
Interaction regions should support occlusions
https://bugs.webkit.org/show_bug.cgi?id=250860
<rdar://103928196>

Reviewed by Tim Horton.

Maintain a separate layer tree for Interaction Regions on top of the
content. This "mirror layer tree" includes layers for occlusions and now
comes with clear ordering rules.

* LayoutTests/interaction-region/click-handler-dynamically-added-expected.txt:
* LayoutTests/interaction-region/click-handler-expected.txt:
* LayoutTests/interaction-region/click-handler-in-shadowed-layer-expected.txt:
* LayoutTests/interaction-region/event-region-overflow-expected.txt:
* LayoutTests/interaction-region/icon-inside-button-single-region-expected.txt:
* LayoutTests/interaction-region/inline-link-dark-background-expected.txt:
* LayoutTests/interaction-region/inline-link-expected.txt:
* LayoutTests/interaction-region/inline-link-in-composited-iframe-expected.txt:
* LayoutTests/interaction-region/inline-link-in-layer-expected.txt:
* LayoutTests/interaction-region/inline-link-in-non-composited-iframe-expected.txt:
* LayoutTests/interaction-region/inline-link-with-pointer-events-none-content-expected.txt:
* LayoutTests/interaction-region/input-type-file-region-expected.txt:
* LayoutTests/interaction-region/input-type-range-region-expected.txt:
* LayoutTests/interaction-region/paused-video-regions-expected.txt:
* LayoutTests/interaction-region/split-inline-link-expected.txt:
* LayoutTests/interaction-region/wrapped-inline-link-expected.txt:
Update expectations with the new dump format.

* LayoutTests/interaction-region/layer-tree-expected.txt: Added.
* LayoutTests/interaction-region/layer-tree.html: Added.
Add test covering the final RemoteLayerTree structure.

* LayoutTests/interaction-region/overlay-expected.txt: Renamed from LayoutTests/interaction-region/inline-link-in-aria-hidden-subtree-expected.txt.
* LayoutTests/interaction-region/overlay.html: Renamed from LayoutTests/interaction-region/inline-link-in-aria-hidden-subtree.html.
Remove the aria-hidden test.
Add test for the occlusions detection.

* LayoutTests/interaction-region/region-area-overflow-does-not-crash-expected.txt:
* LayoutTests/interaction-region/region-area-overflow-does-not-crash.html:
Update test to also include a big occlusion region.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.getCALayerTree):
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
(WTR::UIScriptController::caLayerTreeAsText const):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::caLayerTreeAsText const):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _caLayerTreeAsText]):
(dumpCALayer):
Introduce a new test helper to get the full remote layer hierarchy.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
Remove the aria-hidden check. It was too strict and is not needed
anymore.
Generate regions for elements likely to occlude interactions.
(WebCore::operator<<):
Update the dump format to differentiate interactions and occlusions.

* Source/WebCore/page/InteractionRegion.h:
(WebCore::operator==):
(WebCore::InteractionRegion::encode const):
(WebCore::InteractionRegion::decode):
Add a `type` enum to the InteractionRegion struct to differentiate
interactions and occlusions.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::applyGeometryPropertiesToLayer):
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
Extract the geometry related properties application code to a new
method.
(WebKit::RemoteLayerTreePropertyApplier::applyProperties):
Only update geometry and event region properties for Interaction Regions
layers.
(WebKit::applyInteractionRegionsHierarchyUpdate):
(WebKit::RemoteLayerTreePropertyApplier::applyHierarchyUpdates):
Apply all hierarchy updates to the Interaction Regions mirror layer tree.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateLayerTree):
Keep the Interaction Regions mirror layer tree on top of the content after
each layer tree update.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeInteractionRegionLayers.mm:
(WebKit::isInteractionLayer):
(WebKit::isOcclusionLayer):
(WebKit::isAnyInteractionRegionLayer):
(WebKit::isInteractionRegionLayer): Deleted.
(WebKit::setInteractionRegionOcclusion):
Update utility functions to differentiate interactions and occlusions.
(WebKit::insertInteractionRegionLayersForLayer):
(WebKit::appendInteractionRegionLayersForLayer): Deleted.
Insert the Interaction Regions back at the beginning of the sublayers array.
(WebKit::updateLayersForInteractionRegions):
Maintain the following order when updating Interaction Regions layers:
1. Occlusions always come first (in no particular order).
2. Interactions come next, ordered by area (biggest to smallest).
3. Other siblings come last, in the same order as the content layers they mirror.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.h:
(WebKit::RemoteLayerTreeNode::interactionRegionsLayer const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeNode.mm:
(WebKit::RemoteLayerTreeNode::detachFromParent):
(WebKit::RemoteLayerTreeNode::initializeLayer):
Add an `interactionRegionsLayer` property to the RemoteLayerTreeNode.

Canonical link: https://commits.webkit.org/259504@main
